src/Data/Voting.hs

```diff
@@ -77,24 +102,41 @@ setDelegation from to (Delegations (DelegationMap dmap) (CoDelegationMap coDmap)
                     in coDmap1
     deleteValue d ds = let ds' = Set.delete d ds in if Set.null ds' then Nothing else Just ds'
 
-canVote :: Voter -> DelegationMap -> Bool
-canVote v (DelegationMap dmap) = isNothing $ Map.lookup v dmap
+canVotePure :: Voter -> DelegationMap -> Bool
+canVotePure v (DelegationMap dmap) = isNothing $ Map.lookup v dmap
 
-getDelegators :: Voter -> CoDelegationMap -> [Voter]
-getDelegators v (CoDelegationMap dmap) = fix voters v
+getDelegatorsPure :: Voter -> CoDelegationMap -> [Voter]
+getDelegatorsPure v (CoDelegationMap dmap) = fix voters v
   where
     voters rec v = v : maybe [] (mconcat . map rec) (Set.toList <$> Map.lookup v dmap)
 
-data Votings = Votings {
-        votings :: Map Voter (Voter, Vote)
-    }
+setVotePure :: Voter -> Vote -> Voter -> Votings -> Votings
+setVotePure voter vote delegator (Votings vmap) =
+    Votings $ Map.insert delegator (voter, vote) vmap
 
-setVote :: Voter -> Vote -> [Voter] -> Votings -> Votings
-setVote voter vote delegators (Votings vmap) = Votings $ foldl setAVote vmap delegators
-  where
-    setAVote m d = Map.insert d (voter, vote) m
 
-vote :: Voter -> Vote -> Delegations -> Votings -> Votings
-vote voter vote (Delegations dmap coDmap) votings
-  | canVote voter dmap = setVote voter vote (getDelegators voter coDmap) votings
-  | otherwise          = votings
+-- * monadic api
+
+class Monad m => DelegationM m where
+    setDelegation :: Voter -> Voter -> m ()
+    canVote       :: Voter -> m Bool
+    getDelegators :: Voter -> m [Voter]
+    voteFor       :: Voter -> Vote -> Voter -> m ()
```

Comment:
```haskell
delegateTo ...
isDelegatee ...
isDelegate ...
```

(on second thoughts, `delegatee` is not much better than `delegator`.  undecided on this one.)